src/JsonLdWriter.php

```diff
@@ -215,9 +245,82 @@ protected function extractPathFromUrl($original)
         return str_replace("/", "", $parts["path"]);
     }
 
-    protected function convertImageUrlToBinary(string $url): ?string
+    protected function convertImageToBinary(array|string $value): mixed
     {
-        // Add code to extract and convert image
+        // First check if we have a simple url string, an array of images or an ImageObject.
+        if (is_string($value)) {
+            // Simple String that we can convert.
+            return $this->convertUrlToBinary($value) ?? $value;
+        } elseif (is_array($value)) {
+            // Either an ImageObject or an Array of images/ImageObjects.
+            if (array_key_exists('@type', $value)) {
+                // ImageObject, we need to find out which field is filled.
+                return $this->convertImageObjectUrlToBinary($value);
+            } else {
+                // Array of images, so we just call this method on every element.
+                $images = [];
+                foreach ($value as $image) {
+                    $images[] = $this->convertImageToBinary($image);
+                }
+            }
+        }
+
+        return $value;
+    }
+
+    protected function convertImageObjectUrlToBinary(array $imageObject): array {
+        if (array_key_exists('contentUrl', $imageObject))  {
+            $contentUrl = $imageObject['contentUrl'];
+
+            $binary = $this->convertUrlToBinary($contentUrl);
+
+            $imageObject['contentUrl'] = $binary ?? $contentUrl;
+            return $imageObject;
+        } elseif (array_key_exists('url', $imageObject)) {
+            $url = $imageObject['url'];
+            
+            $binary = $this->convertUrlToBinary($url);
+
+            $imageObject['url'] = $binary ?? $url;
+        }
+
+        return $imageObject;
+    }
+
+    protected function convertUrlToBinary($url): ?string {
+        // Check if the string is a valid URL.
+        if(!filter_var($url, FILTER_VALIDATE_URL)) {
+            return null;
+        }
+
+        // TODO make this configurable.
+        // Set a timeout for the call. 
+        $ctx = stream_context_create(array('http'=>
+            array(
+                'timeout' => $this->fileDownloadTimeoutWindow,
+            )
+        ));
+
+        $content = file_get_contents($url, true, $ctx, 0, $this->fileDownloadSizeLimit);
+
+        if (!$content) {
+            return null;
+        }
+
+        $encodedContent = base64_encode($content);
+
+        $mimeTypeSignatures = [
+            'iVBORw0KGgo'=> 'image/png',
+            '/9j/' => 'image/jpg',
+            'R0lGODdh' => 'image/gif',
+            'R0lGODlh'=> 'image/gif'
+        ];
+
+        foreach ($mimeTypeSignatures as $sig => $mimeType) {
+            if (str_starts_with($encodedContent, $sig)) {
+                return 'data:' . $mimeType . ';base64,' . $encodedContent;
+            }
+        }
 
         return null;
```

Comment:
This looks to me like we are:

a) swallowing unsupported images (replacing URL with null)
b) only supporting some variants of git, jpg and png (there should be more mimetypes for images, see e.g. https://www.php.net/manual/en/function.mime-content-type.php#87856 )

In combination that is not so nice. I suggest two things:

* Support more of the available image mimetypes
* Make sure we are keeping the URL just in case we did not support the mimetype

Reply:
Regarding a):
The URL should not be swallowed, instead, the null return value is ignored using the null coalescing operator (i.e. https://github.com/audriga/html2jsonld-php/blob/77abea8869541113c86ca6602e6852b5038c9b8d/src/JsonLdWriter.php#L253). However, I guess that returning the URL should be less error prone, so I will change that anyway.

Reply:
And regarding b):
I still need to find a list of Base64 file signatures for more types of images. I could also find them out myself but did not get to it yet.

Reply:
a) looks nice! No need to change it.

Reply:
OK lets leave b) open for now.